### PR TITLE
12/add roles

### PR DESCRIPTION
**Pull request by BenAAndrew**

### CODE REVIEW

File: Angular/src/app/add-role-modal/add-role-modal.component.html [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+        *ngIf="submitted && formControls.roleName.errors"
+        class="invalid-feedback"
+      >
+        <div *ngIf="formControls.roleName.errors.required">
+          Role Name is a required field
+        </div>
+      </div>
+    </div>
+
+    <div class="form-group">
```
Comment (Fluffasaurus) [inline Angular/src/app/add-role-modal/add-role-modal.component.html]: Make all fields in this form required, even though some of the current data from the career lattice is missing. These should be required to ensure that empty role information isn't added.

File: Angular/src/app/add-role-modal/add-role-modal.component.html [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+        *ngIf="submitted && formControls.roleSummary.errors"
+        class="invalid-feedback"
+      >
+        <div *ngIf="formControls.roleSummary.errors.maxlength">
+          Max characters reached for Role Summary
+        </div>
+      </div>
+    </div>
+
+    <div class="form-group">
```
Comment (Fluffasaurus) [inline Angular/src/app/add-role-modal/add-role-modal.component.html]: same as above

File: Angular/src/app/add-role-modal/add-role-modal.component.html (hunk 1 above)
Comment (BenAAndrew) [reply Angular/src/app/add-role-modal/add-role-modal.component.html]: sorted

**Comment by jamescrone** [comment]
Fix merge conflicts

File: Angular/src/app/admin-auth-guard/admin-auth-guard.component.ts [hunk 3]
```diff
@@ -0,0 +1,42 @@
+import { Injectable } from "@angular/core";
+import { Router, CanActivate, ActivatedRouteSnapshot, RouterStateSnapshot } from "@angular/router";
+import { CookieService } from 'ngx-cookie-service';
+import { decode, verify } from 'jsonwebtoken'
+import { AuthGuardComponent } from '../auth-guard/auth-guard.component';
+
+@Injectable()
+export class AdminAuthGuardComponent implements CanActivate {
+    constructor(private router: Router, private cookieService: CookieService,
```
Comment (jurassic-marc) [inline Angular/src/app/admin-auth-guard/admin-auth-guard.component.ts]: Could add to new line

File: Angular/src/app/edit-roles/edit-roles.component.html [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+              scope="col"
+              [attr.colspan]="capabilities.length + 1"
+              rowspan="1"
+              *ngFor="let department of departmentName">
+              {{ department.name }}
+            </th>
+          </tr>
+          <tr>
+              <th class='job-family'></th>
+              <th scope="row" *ngFor='let capability of capabilities' class='job-headings'>{{capability.name}}</th>
```
Comment (jurassic-marc) [inline Angular/src/app/edit-roles/edit-roles.component.html]: Split this like you did in career table

`         
          scope="row"
          [class.job-level-inspire]="band.band_id <= 3"
          [class.job-level-explore]="band.band_id > 3 && band.band_id <= 6"
          [class.job-level-discover]="band.band_id > 6"
          class="job-level"
          (click)="this.modalService.openBandInfoModal(band.band_id)">
          {{ band.name }}
`

File: Angular/src/app/landing-page/landing-page.component.html [hunk 5]
```diff
@@ -1,67 +1,63 @@
 <header>
     <app-navbar></app-navbar>
 </header>
-
+<link rel="stylesheet" href="https://maxcdn.bootstrapcdn.com/bootstrap/4.0.0/css/bootstrap.min.css"
+    integrity="sha384-Gn5384xqQ1aoWXA+058RXPxPg6fy4IWvTNh0E263XmFcJlSAwiGgFAW/dAiS6JXm" crossorigin="anonymous">
```
Comment (jurassic-marc) [inline Angular/src/app/landing-page/landing-page.component.html]: Remove this redundant style import

File: Angular/src/app/navbar/navbar.component.html [hunk 6]
```diff
@@ -1,10 +1,10 @@
 <nav class="navbar navbar-dark bg-primary">
     <img src="../../assets/Kainos_Logo.png" style="height: 50px;" alt="Kainos Logo" class="d-inline-block align-top" alt="Kainos logo">
```
Comment (jurassic-marc) [inline Angular/src/app/navbar/navbar.component.html]: Remove duplicate `alt="Kainos Logo"`... my bad

File: Express-API/roles.js [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+const createRoleCapabilityLinkQuery = "INSERT INTO role_capability(role_id,capability_id) VALUES (?,?) "
+
+class RolesHandler {
+    constructor( config ) { }
+
+    createRole(params, db){
+        return new Promise( ( resolve, reject ) => {
+            try {
+                db.query(createRoleQuery,[params.roleName,params.departmentId,params.bandId,params.summary,params.jobSpecUrl]).then(rows => {
+                    console.log(rows.insertId)
```
Comment (jurassic-marc) [inline Express-API/roles.js]: Can this comment be removed?

File: Angular/src/app/add-role-modal/add-role-modal.component.ts [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+  constructor(
+    private formBuilder: FormBuilder,
+    public activeModal: NgbActiveModal,
+    private dataService: DataService,
+  ) {}
+
+  ngOnInit() {
+    this.addRoleForm = this.formBuilder.group({
+      roleName: ["", Validators.maxLength(100)],
+      roleSummary: ["", Validators.maxLength(65000)],
```
Comment (matthewdoreilly) [inline Angular/src/app/add-role-modal/add-role-modal.component.ts]: 65000 characters max for a role summary? Seems a bit much

File: Angular/src/app/admin-auth-guard/admin-auth-guard.component.ts [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+@Injectable()
+export class AdminAuthGuardComponent implements CanActivate {
+    constructor(private router: Router, private cookieService: CookieService, 
+      private authGuard: AuthGuardComponent){}
+
+    cookieName = 'adminMode'
+
+    isAdminMode(){
+      return this.cookieService.get(this.cookieName) == 'true'
+    }
```
Comment (matthewdoreilly) [inline Angular/src/app/admin-auth-guard/admin-auth-guard.component.ts]: I would add return types to your methods, and types to your variables in this file.
E.g., `isAdminMode(): boolean`

File: Express-API/roles.js (hunk 7 above)
Comment (matthewdoreilly) [reply Express-API/roles.js]: Yeah I'd definitely remove that log here

File: Angular/src/app/admin-auth-guard/admin-auth-guard.component.ts (hunk 9 above)
Comment (BenAAndrew) [reply Angular/src/app/admin-auth-guard/admin-auth-guard.component.ts]: sorted

**Pull request merged**